src/awkward1/operations/structure.py

```diff
@@ -432,42 +433,54 @@ def with_field(base, what, where=None, highlevel=True):
     #ak.with_field, so performance is not a factor in choosing one over the
     other.)
     """
-    base = awkward1.operations.convert.to_layout(
-        base, allow_record=True, allow_other=False
-    )
-    if base.numfields < 0:
-        raise ValueError("no tuples or records in array; cannot add a new " "field")
 
-    what = awkward1.operations.convert.to_layout(
-        what, allow_record=True, allow_other=True
-    )
+    if isinstance(where, (list, tuple)) and len(where) > 1:
```

Comment:
`isinstance(where, Iterable) and all(isinstance(x, str) for x in where) and len(where) > 1`

The `Iterable` name is already imported (with a try-catch for Pythons 2.7 and 3.5+).